// appium-android-driver/build/lib/commands/recordscreen.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _asyncbox = require('asyncbox');

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _appiumSupport = require('appium-support');

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var commands = {},
    extensions = {};

var RETRY_PAUSE = 1000;

/**
* record the display of devices running Android 4.4 (API level 19) and higher.
* It records screen activity to an MPEG-4 file. Audio is not recorded with the video file.
*
* @param filePath the video file name
*                 for example, "/sdcard/demo.mp4"
* @param videoSize the format is widthxheight.
*                  if it is "default", the default value is the device's native display resolution (if supported),
*                  1280x720 if not. For best results,
*                  use a size supported by your device's Advanced Video Coding (AVC) encoder.
*                  for example, "1280x720"
* @param timeLimit the maximum recording time, in seconds. if it is -1, the default and maximum value is 180 (3 minutes).
* @param bitRate the video bit rate for the video, in megabits per second.
*                if it is -1, the default value is 4Mbps. You can increase the bit rate to improve video quality,
*                but doing so results in larger movie files.
*                for example, 6000000
*
*/
commands.startRecordingScreen = function callee$0$0(filePath, videoSize, timeLimit, bitRate) {
  var apiLevel, cmd;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    var _this2 = this;

    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!this.isEmulator()) {
          context$1$0.next = 2;
          break;
        }

        throw new Error('Screen recording does not work on emulators');

      case 2:
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(this.adb.getApiLevel());

      case 4:
        apiLevel = context$1$0.sent;

        if (!(apiLevel < 19)) {
          context$1$0.next = 7;
          break;
        }

        throw new Error('Screen recording not available on API Level ' + apiLevel + '. Minimum API Level is 19.');

      case 7:
        context$1$0.next = 9;
        return _regeneratorRuntime.awrap(this.adb.fileExists(filePath));

      case 9:
        if (!context$1$0.sent) {
          context$1$0.next = 11;
          break;
        }

        throw new Error('Screen recording failed: \'' + filePath + '\' already exists.');

      case 11:
        cmd = ['screenrecord', filePath];

        if (_appiumSupport.util.hasValue(videoSize)) {
          cmd.push('--size', videoSize);
        }
        if (_appiumSupport.util.hasValue(timeLimit)) {
          cmd.push('--time-limit', timeLimit);
        }
        if (_appiumSupport.util.hasValue(bitRate)) {
          cmd.push('--bit-rate', bitRate);
        }

        // wrap in a manual Promise so we can handle errors in adb shell operation
        context$1$0.next = 17;
        return _regeneratorRuntime.awrap(new _bluebird2['default'](function callee$1$0(resolve, reject) {
          var err;
          return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
            var _this = this;

            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                err = undefined;

                _logger2['default'].debug('Beginning screen recording with command: \'adb ' + cmd.join(' ') + '\'');
                // do not await here, as the call runs in the background and we check for its product
                this.adb.shell(cmd)['catch'](function (e) {
                  err = e;
                });

                // there is the delay time to start recording the screen, so, wait until it is ready.
                // the ready condition is
                //   1. check the movie file is created
                //   2. check it is started to capture the screen
                context$2$0.prev = 3;
                context$2$0.next = 6;
                return _regeneratorRuntime.awrap((0, _asyncbox.retryInterval)(10, RETRY_PAUSE, function callee$2$0() {
                  var size;
                  return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
                    while (1) switch (context$3$0.prev = context$3$0.next) {
                      case 0:
                        if (!err) {
                          context$3$0.next = 2;
                          break;
                        }

                        return context$3$0.abrupt('return');

                      case 2:
                        size = this.adb.fileSize(filePath);

                        if (!(size <= 32)) {
                          context$3$0.next = 5;
                          break;
                        }

                        throw new Error('Remote file \'' + filePath + '\' found but it is still too small: ' + size + ' bytes');

                      case 5:
                      case 'end':
                        return context$3$0.stop();
                    }
                  }, null, _this);
                }));

              case 6:
                context$2$0.next = 11;
                break;

              case 8:
                context$2$0.prev = 8;
                context$2$0.t0 = context$2$0['catch'](3);

                err = context$2$0.t0;

              case 11:
                if (!err) {
                  context$2$0.next = 14;
                  break;
                }

                _logger2['default'].error('Error recording screen: err.message');
                return context$2$0.abrupt('return', reject(err));

              case 14:
                resolve();

              case 15:
              case 'end':
                return context$2$0.stop();
            }
          }, null, _this2, [[3, 8]]);
        }));

      case 17:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 18:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

/**
* stop recording the screen.
*/
commands.stopRecordingScreen = function callee$0$0() {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.prev = 0;
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(this.adb.killProcessesByName('screenrecord'));

      case 3:
        context$1$0.next = 8;
        break;

      case 5:
        context$1$0.prev = 5;
        context$1$0.t0 = context$1$0['catch'](0);

        _logger2['default'].errorAndThrow('Unable to stop screen recording: ' + context$1$0.t0.message);

      case 8:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[0, 5]]);
};

_Object$assign(extensions, commands);
exports.commands = commands;
exports['default'] = extensions;

// this function is suppported on the device running android 4.4(api level 19)

//if there's same file in the path, then thorws error

//make adb command
// eslint-disable-line curly
